File: 83/cpp/main.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        // Store the origin node, we will return this later
        ListNode* tail = head;

        while (tail != nullptr && tail->next != nullptr) {
            // If the current head and next head have the same values, keep
            // going until we find a different value
            if (tail->val == tail->next->val) {
                tail->next = tail->next->next;
            } else {
                tail = tail->next;
            }
        }

        return head;
    }
};

// Helper function to print the linked list
void printList(ListNode* head) {
    while (head != nullptr) {
        std::cout << head->val << " ";
        head = head->next;
    }
    std::cout << std::endl;
}

int main() {
    Solution solution;

    // Test Case 1: head = [1, 1, 2]
    ListNode* head1 = new ListNode(1, new ListNode(1, new ListNode(2)));
    std::cout << "Original List 1: ";
    printList(head1);
    ListNode* deduplicated1 = solution.deleteDuplicates(head1);
    std::cout << "Deduplicated List 1: ";
    printList(deduplicated1);

    // Test Case 2: head = [1, 1, 2, 3, 3]
    ListNode* head2 = new ListNode(1, new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3)))));
    std::cout << "Original List 2: ";
    printList(head2);
    ListNode* deduplicated2 = solution.deleteDuplicates(head2);
    std::cout << "Deduplicated List 2: ";
    printList(deduplicated2);

    // Test Case: head = [0, 0, 0, 0, 0]
    ListNode* head3 = new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0)))));
    std::cout << "Original List 3: ";
    printList(head3);
    ListNode* deduplicated3 = solution.deleteDuplicates(head3);
    std::cout << "List after removing duplicates: ";
    printList(deduplicated3);

    return 0;
}

File: 83/python3/main.py
#!/usr/bin/env python3

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        tail = head

        while (tail != None and tail.next != None):
            if (tail.val == tail.next.val):
                tail.next = tail.next.next
            else:
                tail = tail.next

        return head

# Helper function to convert a Python list into a linked list
def build_linked_list(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for val in arr[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

# Helper function to convert a linked list back to a Python list (for easy verification)
def linked_list_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

head = build_linked_list([1,1,2])
sol = Solution()
merged = sol.deleteDuplicates(head)
print("Merged list:", linked_list_to_list(merged))

head = build_linked_list([1,1,2,3,3])
sol = Solution()
merged = sol.deleteDuplicates(head)
print("Merged list:", linked_list_to_list(merged))

head = build_linked_list([0,0,0,0,0,0])
sol = Solution()
merged = sol.deleteDuplicates(head)
print("Merged list:", linked_list_to_list(merged))
